### PR TITLE
Faster sir decoding

### PR DESCRIPTION
**Pull request by bjorn3**

### CODE REVIEW

File: yktrace/src/hwt/mapper.rs [hunk 1]
```diff
@@ -94,17 +94,18 @@ fn extract_labels() -> Result<Vec<(u64, (String, u32))>, gimli::Error> {
     // Extract labels
     let mut labels = Vec::new();
 
-    let loader = |id: gimli::SectionId| -> Result<borrow::Cow<[u8]>, gimli::Error> {
+    let loader = |id: gimli::SectionId| -> Result<&[u8], gimli::Error> {
         Ok(object
-            .section_data_by_name(id.name())
-            .unwrap_or(borrow::Cow::Borrowed(&[][..])))
+            .section_by_name(id.name())
+            .map(|sec| sec.data().expect("failed to decompress section"))
```
Comment (vext01) [inline yktrace/src/hwt/mapper.rs:100]: What impact does compression have on performance?

File: yktrace/src/sir.rs [hunk 2]
```diff
@@ -64,7 +65,9 @@ lazy_static! {
 
 impl Sir {
     pub fn read_file(file: &Path) -> Result<Sir, ()> {
-        let ef = elf::File::open_path(file).unwrap();
+        // SAFETY: Not really, we hope that nobody changes the file underneath our feet.
+        let data = unsafe { Mmap::map(&File::open(file).unwrap()).unwrap() };
```
Comment (vext01) [inline yktrace/src/sir.rs:69]: So mmap is faster?

**Comment by vext01** [comment]
Thanks for this.

How much time does all of this shave off?

File: yktrace/src/hwt/mapper.rs (hunk 1 above)
Comment (bjorn3) [reply yktrace/src/hwt/mapper.rs:100]: I don't know. ykrustc is not emitting compressed elf sections. They are an ELF extension that is as far as I know only used for debug info in some cases.

File: yktrace/src/sir.rs (hunk 2 above)
Comment (bjorn3) [reply yktrace/src/sir.rs:69]: There are two options when using `object`. Either you load the whole file into a `Vec<u8>` or you mmap the file. The former would also load the `.text` and `.data` sections, not just the `.yksir_*` sections. Rustc also uses mmap for metadata loading. I have not benchmarked both approaches though.

**Comment by bjorn3** [comment]
In debug mode:

```
Benchmark #1: ./ykview_master_debug ./yktrace_test_master
  Time (mean ± σ):      8.852 s ±  0.094 s    [User: 8.636 s, System: 0.216 s]
  Range (min … max):    8.777 s …  9.103 s    10 runs
 
Benchmark #2: ./ykview_faster_sir_decoding_debug ./yktrace_test_faster_sir_decoding
  Time (mean ± σ):      6.552 s ±  0.010 s    [User: 6.435 s, System: 0.117 s]
  Range (min … max):    6.536 s …  6.564 s    10 runs
 
Summary
  './ykview_faster_sir_decoding_debug ./yktrace_test_faster_sir_decoding' ran
    1.35 ± 0.01 times faster than './ykview_master_debug ./yktrace_test_master'
```

File size:

```
-rwxr-xr-x 1 bjorn3 bjorn3  84M Sep  5 17:14 yktrace_test_faster_sir_decoding
-rwxr-xr-x 1 bjorn3 bjorn3 110M Sep  5 16:55 yktrace_test_master
```

It is faster and produces smaller SIR.

<details><summary>Bonus benchmark</summary>

In release mode:

```
Benchmark #1: ./ykview_master ./yktrace_test_master
  Time (mean ± σ):      1.043 s ±  0.013 s    [User: 847.8 ms, System: 194.7 ms]
  Range (min … max):    1.035 s …  1.093 s    20 runs
 
  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet PC without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.
 
Benchmark #2: ./ykview_faster_sir_decoding ./yktrace_test_faster_sir_decoding
  Time (mean ± σ):     666.1 ms ±   1.7 ms    [User: 558.0 ms, System: 108.0 ms]
  Range (min … max):   663.7 ms … 669.6 ms    20 runs
 
Summary
  './ykview_faster_sir_decoding ./yktrace_test_faster_sir_decoding' ran
    1.57 ± 0.02 times faster than './ykview_master ./yktrace_test_master'
```

</details>

File: yktrace/src/sir.rs [hunk 3]
```diff
@@ -9,9 +8,11 @@ use std::{
     fmt::{self, Debug, Display, Write},
     io::Cursor,
     iter::Iterator,
-    path::Path
+    path::Path, fs::File
 };
 use ykpack::{self, bodyflags, Body, CguHash, Decoder, Local, Pack, Ty};
+use memmap::Mmap;
+use object::{ObjectSection, Object}; // FIXME kill.
```
Comment (vext01) [inline yktrace/src/sir.rs:15]: What's the FIXME about?

**Comment by vext01** [comment]
I wish we could build just the SIR loading stuff with optimisations. We never need to trace that so we can optimise to our heart's content!

File: yktrace/src/sir.rs (hunk 3 above)
Comment (bjorn3) [reply yktrace/src/sir.rs:15]: Huh, I can't remember writing that.

**Comment by vext01** [comment]
Related:
https://github.com/rust-lang/rust/issues/54882

**Comment by bjorn3** [comment]
Would it be possible to build everything except for ykrt without tracing and then make ykrt call into the rest of the crates using `#[do_not_trace]` wrappers when necessary?

File: yktrace/src/sir.rs (hunk 3 above)
Comment (bjorn3) [reply yktrace/src/sir.rs]: What is even weirder is that that line was created using the rust-analyzer import assist as far as I know.

**Comment by vext01** [comment]
(I'm debugging a buildbot problem, please ignore these `bors try`s)

bors try

**Comment by vext01** [comment]
> Would it be possible to build everything except for ykrt without tracing and then make ykrt call into the rest of the crates using #[do_not_trace] wrappers when necessary?

In theory yes, I guess we could force the optimisation flags in those crates (in Cargo.toml)?

**Comment by vext01** [comment]
Please squash.

**Comment by bjorn3** [comment]
Splat

**Comment by bjorn3** [comment]
Fixed rustfmt. This will need a ykrustc PR too because of the switch to bincode.

**Comment by vext01** [comment]
> Fixed rustfmt. This will need a ykrustc PR too because of the switch to bincode.

Do you just need a rebuild of ykrustc, or are there code changes?

**Comment by bjorn3** [comment]
Just a rebuild against the new ykpack. Opened as https://github.com/softdevteam/ykrustc/pull/128. Can you check that I made all changes necessary for cycle breaking?

**Comment by bjorn3** [comment]
Wait, I need to fix a warning in this PR.

**Comment by vext01** [comment]
Can we wait for the upstream sync to go in before merging this?

I'm just stuck with:
https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/path.20trimming.20problem

**Comment by bjorn3** [comment]
The ykrustc PR already got a `bors try`, so the cached rustc already needs this PR.

**Comment by vext01** [comment]
OK. I'll probably end up having to redo the sync. Rebasing pretty much never works well for them.

bors r+

**Pull request merged**